### ISSUE TITLE
NPE in InvalidObservationException.HistoryElement

### ISSUE THREAD

**Issue by marcantoine-bibeau** [micrometer-metrics/micrometer#5702]
**Describe the bug**
NPE in InvalidObservationException.HistoryElement
```
StackTraceElement[] currentStackTrace = Thread.getAllStackTraces().get(Thread.currentThread()); 
```
returns null and 
this.stackTrace = findRelevantStackTraceElements(currentStackTrace) do
```
stackTrace.length
```

That throws a NPE

**Environment**
 - Micrometer version: 1.14.1
 - Micrometer registry: otel

**Comment by jonatan-ivanov** [comment]
Thank for the issue!
Since `getAllStackTraces` returns a `Map`, theoretically, it is possible that `get` will return `null`. But I have no idea how could that happen in practice. Could you please provide a minimal sample project to reproduce this issue so we can more easily investigate and ensure any fix is working properly for your use case?

**Comment by marcantoine-bibeau** [comment]
I would like to provide a sample project but I'm not able to reproduce in small project :( I'm working on an existing project (that is not small ;)), I wanted to update to Spring 3.4 and updating dependancies but I'm getting that error... Will be very hard to give you more :(

**Comment by jonatan-ivanov** [comment]
Could you please at least tell us some details about your use-case and environment?
- I assume you are using this for tests, are those unit tests or some kind of integration tests when you start the whole app? 
- Are you using JUnit or something else? Are you also using PowerMock?
- Is this intermittent or happens all the time?
- Is this also happening if you put a `Thread.sleep(3_000)` to the last line of your test?
- Is this happening when you instrument your own code or when a third-party lib/framework instruments itself?
- Which JVM (version/vendor, is it "default" hotspot or OpenJ9/Android/Azul Prime/etc.)?
- Are you using virtual threads?
- Are you using WebFlux or any reactive library (Reactor)?
- Are you using any Java agents?

Is it possible that you start removing things from the big project until the issue disappears and you could tell us what was that change?

**Comment by marcantoine-bibeau** [comment]
- These are End-2-End Tests using TestContainers
- Using Jupiter. Not Using Power, a bit of Mockito
- Happening all the time 
- I can try 
- That is my own code instrumentation
- Amazon Correto 22
- Virtual Thread all the way 
- Not using Reactor, Using Http-Component from Apache
- Not using any agent

I can ;)

**Comment by jonatan-ivanov** [comment]
Thank you for the quick answers!
I have a few more:
- Is your instrumentation uses `@Observed` or you create an Observation manually? (I guess manually but want to confirm.)
- Btw the fact that this is throwing an NPE also indicates that there is a problem with the instrumentation
- Are you creating virtual threads on your own?

Is it possible that you can show us your instrumentation in a way that it won't reveal any company-specific details?
For example, instead of this: `myCompanyCarServiceHttpClient.getSecretCarInformation(...)` you could modify it to something like: `httpsClient.getSomething(...)`, or instead of `observation.lowCardinalityKeyValues("superSecretField", response.headers().get("superSecretField"))` something like: `observation.lowCardinalityKeyValues("testField", response.headers().get("testField"))`.

What I'm interested in is:
- How the observation is created, interacted with
- Is there any interesting detail around it (threading, locking, synchronized, JNI, reflection, summoning daemons in the language of Mordor)?

**Comment by marcantoine-bibeau** [comment]
- I use @Observed, I even have AOP that is where its becoming a bit more complexe, but I'm not doing nothing crazy ;)
- I do not create my virtual thread manually

I'll do my best to start investigating more, and give you more details I hope today or tomorrow

**Comment by jonatan-ivanov** [comment]
Oh wait so you have a custom aspect?
Because using `@Observed` with the default `ObservedAspect` should not trigger any `InvalidObservationException` (which triggers the NPE).

**Comment by marcantoine-bibeau** [comment]
I think I may have found something that I hope could help here.
```
HistoryElement(EventName eventName) {
            this.eventName = eventName;
            StackTraceElement[] currentStackTrace = Thread.getAllStackTraces().get(Thread.currentThread());
            this.stackTrace = findRelevantStackTraceElements(currentStackTrace);
        }
```
at line 
```
StackTraceElement[] currentStackTrace = Thread.getAllStackTraces().get(Thread.currentThread());
```
```
Thread.currentThread()
```
is a VirtualThread `VirtualThread[#244,AuthenticationTokenRunnable]/runnable@ForkJoinPool-2-worker-8`

while `Thread.getAllStackTraces()` does not contains any VirtualThread 
if I do `Thread.getAllStackTraces().get(((VirtualThread) Thread.currentThread()).carrierThread)` It returns an Thread Object. 

With that said, I'm not super knowledgeable regarding VirtualThread implementation so please let me know if that info helps or not ;)

**Comment by marcantoine-bibeau** [comment]
Just saw your commit and yes I agree with your comment! Thanks a bunch for your help and quick response and fix :)

**Comment by jonatan-ivanov** [comment]
It does, thank you!
Could you please also answer my question (https://github.com/micrometer-metrics/micrometer/issues/5702#issuecomment-2501936132) above:
>Oh wait so you have a custom aspect?
Because using @Observed with the default ObservedAspect should not trigger any InvalidObservationException (which triggers the NPE).

The "fix" should be released as a snapshot (`1.14.0-SNAPSHOT`) in a few minutes, could you please check if it fixes the issue for you? (see how to use snapshots in the readme.)

**Comment by marcantoine-bibeau** [comment]
> Oh wait so you have a custom aspect? Because using `@Observed` with the default `ObservedAspect` should not trigger any `InvalidObservationException` (which triggers the NPE).

I have an @Aspect that does
```
private boolean haveObservedAnnotation(Method method) {
    if (method.getDeclaringClass().isAnnotationPresent(Observed.class)) {
      return true;
    }

    return method.getAnnotation(Observed.class) != null;
  }
```

But that was not throwing any exception. Now I see the Updated comments on `io.micrometer.observation.aop.ObservedAspect` that allows me to augment my observation. I will definitely llok into to adapt to newest micrometer version. That I implemented when Springboot 3.0 came out ;)

**Comment by jonatan-ivanov** [comment]
I was able to repro the issue and verify the fix. This results in an NPE while the line that is commented out does not:
```java
Thread thread = Thread.ofVirtual().start(() -> {
	StackTraceElement[] currentStackTrace = Thread.getAllStackTraces().get(Thread.currentThread());
	//StackTraceElement[] currentStackTrace = Thread.currentThread().getStackTrace();
	Arrays.stream(currentStackTrace).forEach(System.out::println);
});
thread.join();
```
